### PR TITLE
Retrieve final qualifier from bytecode methods [TG-2893]

### PR DESCRIPTION
**Pull request by allredj**

### CODE REVIEW

File: jbmc/src/java_bytecode/java_types.h [hunk 1]
```diff
@@ -161,12 +161,12 @@ class java_class_typet:public class_typet
     return set(ID_is_anonymous, is_anonymous_class);
   }
 
-  bool get_final()
+  const bool get_is_final() const
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_types.h]: `const` before `bool` has no effect here.

**Review by mgudemann — APPROVED** [review]
LGTM

**Review by thk123 — COMMENTED** [review]
LGTM except tests :) Please write a couple of unit tests for this:

 - static final method
 - non-static final method 

They can be added to jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp

**Comment by allredj** [comment]
@thk123 That's not testable because we don't actually use that value in cbmc. I'm retrieving the qualifier because we will certainly need it later in test-gen.
What kind of tests were you thinking about?

**Comment by mgudemann** [comment]
@allredj I think it could be unit tested

**Comment by allredj** [comment]
Unit testing getters and setters seems like a waste of time to me. Any meaningful test should at least invoke the parser. Any ideas on how to do that?

**Comment by tautschnig** [comment]
> Unit testing getters and setters seems like a waste of time to me. Any meaningful test should at least invoke the parser. Any ideas on how to do that?

Just on the first part (it may be necessary that this requires parsing, I don't know): you are effectively building a (very partial) specification by writing those unit tests. The following use cases seem relevant and possibly non-trivial:
1) Run the getter without having used the setter before - what is the expected result?
2) Run the setter and then the getter, and expect the result to be what the setter said.
3) Run the setter again (with a different value), and expect that the getter indeed returns the updated value.

**Comment by thk123** [comment]
For the avoidance of doubt - I have spoken with @allredj out of band - he is going to add unit test similar to the one I linked to.

**Comment by allredj** [comment]
@thk123 added a couple of unit tests to check that the final qualifier for methods is indeed obtainable through `get_is_final`

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 76a8f8ce7abe70b4354c44563de89ccf6fd4996b).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 088d10106a560c75cee99325a8f1c69135c885c5).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 2]
```diff
@@ -57,3 +57,37 @@ SCENARIO(
...
+    }
+    WHEN("When parsing a non-final method")
+    {
+      const symbolt function_symbol =
+        symbol_table.lookup_ref("java::ClassWithFinalMethod.nonFinalFunc:()I");
+      const java_method_typet &function_type =
+        require_type::require_java_method(function_symbol.type);
+      THEN("The method should not be marked as final")
+      {
+        REQUIRE(!function_type.get_is_final());
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:89]: :pick: `REQUIRE_FALSE` is a bit clearer than `REQUIRE(!` (@allredj has said there would be a follow up PR - this is my only comment about the tests!)

**Pull request merged**